### ISSUE TITLE
Some features broken after commit

### ISSUE THREAD

**Issue by ghost** [clearkimura/Customizer#80]
Latest commit 4fe0a00 cannot run **Terminal**, **Archive** and **Desktop** via GUI.

The last devel branch I had tested working well was commit 53a1566 to test potential fix for issue #78 which means, one of the commits between these two points had broke those features.

I didn't test via CLI alone, but I ran GUI from Terminal so I could see output there saying something like, "filesystem invalid or missing" when trying to run **Desktop** feature. It looks like it failed to look for xsession file directory or something.

As for other two features, these windows are too fast to disappear and the error may have not appeared any, or may be something similar to above (my memory is fuzzy, can't recall totally).

This issue is applicable to any ISO image loaded to latest devel branch commit 4fe0a00. In my case, I have tested with both 12.04 (contains minimal Xfce) and 14.04 (contains only minimal CLI) ISO images that were working well as commit 53a1566.

I also reinstalled commit 53a1566 to confirm that 12.04 and 14.04 ISO images are working well.

I browse the codes for all the commits between 4fe0a00 and 53a1566, but I couldn't identify which commit broke the features. My best guess is, either 5c1859f or 4fe0a00 itself.

P.S.:  I couldn't upload the detailed information and log, due to Internet connection downtime following thunderstorms hit my hometown area

P.S.S.: I suppose I could have tried download the code for all commits in between, install and test it one by one, but I really had limited time for to check out the situation. Forgive me.

**Comment by fluxer** [comment]
This sounds really like a corrupted filesystem, the common check verifies that 'bin', 'sbin', 'usr/bin', 'usr/sbin', 'etc', 'lib', 'usr/lib' are present in the filesystem (e.g. /home/FileSystem/bin, /home/FileSystem/sbin, etc.) which would be the first indicator that something has gone wrong (e.g. unexpected power cut during extraction without proper cleanup). Thats for the CLI. In addition the GUI checks for additional case where if etc/caspe.conf is not present in the FileSystem directory ,which is neededed for certain functionality (the hostname and live user), the copy will be considered as corrupted disabling certain functionality and pop an error message that clean is recommended and the clean button itself will be enabled. You report sounds like a mixed bag of these but casper.conf does exists and proves that no existing check is sufficient enough.

> As for other two features, these windows are too fast to disappear and the error may have not appeared any, or may be something similar to above (my memory is fuzzy, can't recall totally).

Indeed, Konsole has the feature to stay on hold until the user closes it but I have no idea if other terminal-emulators have such feature so as I've explained (I think) in one of the commits - the backend (CLI) can raise exceptions for the GUI _or_ (something we two have talked about) Customizer must be focused on the GUI interface only.

So, course of action - I recommend you to use the plain python script files and check if it does work for you, meaning no to build and install it via `make`. From the _src_ directory run one of the following:
    sudo python2 main.py -D -c

The above command will show you which directory does not exists in cyan color in the CLI filesystem check.

```
kdesu -t python2 gui.py
gksu python2 gui.py
```

The above command will ensure that it's not a bug related to Nuitka. In addition you can build the executables with the host Nuitka (you will have to install it first if you have not already) likes this:
make NUITKA=nuitka && sudo make install

Lastly, check manually for the presence and delete /home/FileSystem and /home/ISO. Adjust paths to your preferences which you set last time. If you are lazy or just scared of the fancy stuff above do this first.

I've tested the latest commit with minor change to use `mount` from busybox and it works as expected for me. Cheers!

**Comment by fluxer** [comment]
Is anyone else experiencing this issue?

**Comment by ghost** [comment]
@fluxer 
First of all, I must inform that the plain python scripts works fine, comments as below.

> From the src directory run one of the following:
> sudo python2 main.py -D -c

Without any ISO image loaded into Customizer, the output was obvious:

```
* Chrooting...
=> Checking
=> Non-existing path: /home/FileSystem/bin
=> Filesystem is missing or corrupted
```

After loading ISO image file to Customizer, the output was showing successful chroot:

```
* Chrooting...
=> Checking
=> Chrooting
root@HOSTNAME:/#
```

> kdesu -t python2 gui.py
> gksu python2 gui.py
> The above command will ensure that it's not a bug related to Nuitka. 

In my case, Xubuntu 14.04 LTS is using `sudo` instead of `gksu` (apparently `gksu` is not recommended in newer releases for some reasons). Anyway, the output--GUI component runs and showing up without any problem at all.

Actually, I think I found the cause of this issue.. could have been related to issue #79 that has been fixed previously. More details will be posted later..

**Comment by fluxer** [comment]
> First of all, I must inform that the plain python scripts works fine, comments as below.

Sounds good, there is something that I've noticed recently - when binaries/libraries produced by Nuitka are stripped sometimes they do not work as expected. If in the end it turns out that it's only issue with them and not the plain Python files can you try to install without stripping them? Here is how to do that:

`make STRIP=echo && sudo make install`

> Actually, I think I found the cause of this issue.. could have been related to issue #79 that has been fixed previously. More details will be posted later..

Ok.

**Comment by ghost** [comment]
@fluxer 
I will put my findings on-hold for now; Let's hear from your side first.

> [..] when binaries/libraries produced by Nuitka are stripped sometimes they do not work as expected. 

This "they do not work as expected" is a problem affecting since first release of devel branch? Could you explain further, as I am not sure what you are facing here (does other user experience this?).

> If in the end it turns out that it's only issue with them and not the plain Python files can you try to install without stripping them?  Here is how to do that:
> make STRIP=echo && sudo make install

Okay, I installed latest devel branch (commit 92e4b60) using above command. I don't see any difference.. What else can I check?

**Comment by fluxer** [comment]
> This "they do not work as expected" is a problem affecting since first release of devel branch? Could you explain further, as I am not sure what you are facing here (does other user experience this?).

It is something I have experienced with other projects that use Nuitka. I have faced it with previous versions of Nuitka (0.4.x.x, not sure which one) but it go away and now this bug seems to be back on the latest version (0.5.x.x), not sure since when. Basicly, libraries and binaries sometimes do not behave the same way as the Python scripts in rare cases. I've noticed the bug only when I strip the binaries with `strip --strip-all` and the libraries with `strip --strip-unneeded` which is a common practice in package managers and the like. If only `strip` (no arguments) is used then they work fine as far as I can tell. And since Nuitka versions 0.5.x.x are introducing a lot of refactoring this is likely to be fixed in future releases.

Come to think of it, we may drop the use of Nuitka. I mean, Customizer does nothing performance critical and Nuitka is only used to reduce the files count and pack all libraries and main executables into a single binary (ok, two with the GUI). The same can be done with Cython but if there may be issues with it too. All-in-all, I'm starting to lean on using plain Python scripts for the project but that's just me, comments appreciated.

> What else can I check?

I have recommended you to check how the compiled binaries behave when not stripped (see my comment above). If they work fine unstripped, I can easily remove the stripping feature in the build system but further test will have to be made with Customizer installed from a deb package as dpkg applies some sort of stripping I guess and it may cause the bug to occure. In that case I can look into the deb building system for options not to strip them and adjust the debian files.

@clearkimura
I'm sorry for asking you constantly to test stuff for me, I do not like Debian/Ubuntu anymore and I'm not using them which is why I can hardly reproduce many of the issues.

Cheers!

**Comment by ghost** [comment]
@fluxer 

> All-in-all, I'm starting to lean on using plain Python scripts for the project but that's just me, comments appreciated.

Just sharing my story: there was once I tempted to use `ember.js` and such library to create Web application (I was excited to play with their demo, with just few lines of code). Yet, because I haven't fully master the language itself, I decided not to jump into "faster" way and was held back by advice.

Whether `ember.js`, `Nuitka` or such, it is definitely useful for our convenience. If I were to voice my opinion, I prefer to use the plain scripts for start. As I get to know what I need, I may start to rely on such libraries and tools to reduce working time, but _not_ solely for convenience.

Despite my saying, consider with a pinch of salt. Alas, one decide what is best for own.

> I'm sorry for asking you constantly to test stuff for me [..]

I do the testing while I still have access to relevant resources with me (the second hand test machine, Ubuntu-based system, some time for my remix experiments). Most importantly, I still continue to use Customizer now. Customizer continues to "alive" because volunteers still exist (even only few of us).

There is no reason for you to say sorry; Express your gratitude instead. Other users, including me, would probably feel better that way. Personally, I am doing this for my own good. Experience counts and likely helpful in the future.

Now that things have calmed down on my side, I can proceed to report my findings.

Not quite, actually. Before reporting about this issue, I shall report other two findings which have higher priority than above. Issues will be posted soon.

**Comment by fluxer** [comment]
@clearkimura 

> Despite my saying, consider with a pinch of salt. Alas, one decide what is best for own.

I will, since I use Nuitka in a lot of my projects I watch as close as possible how it evolves. I will evaluate further if Customizer should use it after the next few releases of it. I do hope it improves and the bug we are facing with it is resolved.

> There is no reason for you to say sorry; Express your gratitude instead. Other users, including me, would probably feel better that way. Personally, I am doing this for my own good. Experience counts and likely helpful in the future.

I shall note that. ;)

**Comment by ghost** [comment]
To summarize, this issue has been addressed in several other issues.

At time of writing, issue #81 and #82 have been fixed and only issue #79 is under "pending" status. This issue will be closed, after resolving the last remaining issue.

**Comment by ghost** [comment]
Now that issue #79 is resolved, there is no other known issues related to this already.
Hence, this issue will be closed.
